app/src/androidTest/java/net/osmtracker/layouts/RepositorySettingsDialogTest.java

```diff
@@ -0,0 +1,112 @@
+import static android.support.test.espresso.matcher.ViewMatchers.withText;
+import static net.osmtracker.util.TestUtils.checkToastIsShownWith;
+import static net.osmtracker.util.TestUtils.getStringResource;
+import static org.hamcrest.core.IsNot.not;
+
+public class RepositorySettingsDialogTest {
+
+    @Rule
+    public ActivityTestRule<AvailableLayouts> mRule = new ActivityTestRule<>(AvailableLayouts.class);
+
+
+    @Test
+    public void testToggleBehaviour(){
+        onView(withId(R.id.github_config)).perform(click());
+
+        onView(withId(R.id.default_server)).perform(click());
+        checkStateAfterToggle(R.id.default_server, R.id.custom_server);
+        checkTextFieldsState(not(isEnabled()));
+        checkTextFieldsDefaultValues();
+
+        onView(withId(R.id.custom_server)).perform(click());
+        checkStateAfterToggle(R.id.custom_server, R.id.default_server);
+        checkTextFieldsState(isEnabled());
+    }
+
+    @Test
+    public void testRepositoryValidation(){
+        checkRepositoryValidity("labexp","osmtracker-android-layouts","master", true);
+        checkRepositoryValidity("labexp","osmtracker-android-layouts","NONE", false);
+    }
+
+
+    public void checkStateAfterToggle(int expectedActiveId, int expectedInactiveId){
+        onView(withId(expectedActiveId)).check(matches(not(isEnabled())));
+        onView(withId(expectedActiveId)).check(matches(isChecked()));
+        onView(withId(expectedInactiveId)).check(matches(not(isChecked())));
+        onView(withId(expectedInactiveId)).check(matches(isEnabled()));
+    }
+
+    public void checkRepositoryValidity(String user, String repo, String branch, boolean isValid){
+        onView(withId(R.id.github_config)).perform(click());
+
+        onView(withId(R.id.custom_server)).perform(click());
+
+        onView(withId(R.id.github_username)).perform(clearText(), typeText(user));
+        onView(withId(R.id.repository_name)).perform(clearText(), typeText(repo));
+        onView(withId(R.id.branch_name)).perform(clearText(), typeText(branch));
+
+        onView(withText(getStringResource(R.string.menu_save))).perform(click());
+
+        String expectedMessage = (isValid) ? getStringResource(R.string.github_repository_settings_valid_server) :
+                getStringResource(R.string.github_repository_settings_invalid_server);
+
+        checkToastIsShownWith(expectedMessage);
+
+        ViewAssertion expectedDialogState = (isValid) ? doesNotExist() : matches(isDisplayed());
+        checkDialogState(expectedDialogState);
+
+    }
+
+    /**
+     * Check if the dialog is shown by looking for its title on the screen
+     */
+    private void checkDialogState(ViewAssertion assertion) {
+        onView(withText(getStringResource(R.string.prefs_ui_github_repository_settings))).check(assertion);
+    }
+
+    /**
+     * Check that the text fields values match the expected default ones
+     */
+    private void checkTextFieldsDefaultValues() {
```

Comment:
The EditText elements for github_username, repository_name and branch_name are always displayed, no matter if the custom repo or default server are selected. The property that changes when the server also changes is if they are enabled of not. So using `isDisplay()` in `check(matches()` seems not useful, I suggest to use `isEnable()`.

Also this function could received a param `Matcher matcher`, and a new call could be added after line 47 with a `isEnabled()` param. (In a similar way its done with `checkTextFieldsState()`

Reply:
Oh ok, but those lines are checking that the OSMTracker.Preferences.VAL_X  strings are shown in the UI, not checking that the EditTexts themselves are enabled. 

For example when it says `onView(withText(OSMTracker.Preferences.VAL_GITHUB_USERNAME)).` it's checking that a view containing VAL_GITHUB_USERNAME is displayed and that only happens if the EditText value is such.